### PR TITLE
feat: resend payload of reoccuring client request

### LINKED ISSUE #20506: Server to resend a response message that client waits for
When server receives the same message id and hash as latest handled server should resend the previous payload to the client.

- Store the latest handled payload until next id is handled
- Only resend payload if client id & hash is the latest handled message

This should help to avoid unnecessary resynchronisations and loggings like:
```
com.vaadin.flow.server.communication.ServerRpcHandler   | Ignoring old duplicate message from the client. Expected: 1748, got: 1747.
```

### PR DESCRIPTION
**Pull request by caalador**
If the client request contains
the previous id and exactly the
same message content respond with
the same payload as previously.

Closes #20506

### CODE REVIEW

File: flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java [hunk 1]
```diff
@@ -215,6 +215,8 @@ public List<Object> getParameters() {
 
     private byte[] lastProcessedMessageHash = null;
 
+    private String lastRequestResponse = "for(;;);[{}]";
```
Comment (tepi) [inline flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java:218]: Is the initial value needed for this field?

File: flow-server/src/main/java/com/vaadin/flow/server/communication/ServerRpcHandler.java [hunk 2]
```diff
@@ -236,6 +236,19 @@ public ResynchronizationRequiredException() {
         }
     }
 
+    /**
+     * Exception thrown then the client side re-sent the same request.
```
Comment (tepi) [inline flow-server/src/main/java/com/vaadin/flow/server/communication/ServerRpcHandler.java:240]: -> "`Exception thrown when the client side re-sends the same request`"

File: flow-server/src/main/java/com/vaadin/flow/server/communication/ServerRpcHandler.java [hunk 3]
```diff
@@ -236,6 +236,19 @@ public ResynchronizationRequiredException() {
...
+     */
+    public static class ResendPayloadException extends RuntimeException {
```
Comment (tepi) [inline flow-server/src/main/java/com/vaadin/flow/server/communication/ServerRpcHandler.java:242]: `PayloadResentException` might be better name, or maybe `ClientResentPayloadException` to emphasize that it is the client doing the resending

**Comment by mcollovati** [comment]
Tested with a custom buggy UIDLRequestHandler that seldom drops a response, returning a 410 response code.
The change proved to work very well. When a response for an UIDL message does not reach the client, then the connection status indicator appears for a moment, but then when the request is re-sent, the client receives the correct response and all proceeds correctly.

File: flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java (hunk 1 above)
Comment (caalador) [reply flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java:218]: I would expect that not really. The thought was more that if something goes wrong the payload can be handled by the client.

File: flow-server/src/main/java/com/vaadin/flow/server/communication/ServerRpcHandler.java (hunk 3 above)
Comment (caalador) [reply flow-server/src/main/java/com/vaadin/flow/server/communication/ServerRpcHandler.java]: done

File: flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java (hunk 1 above)
Comment (caalador) [reply flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java]: removed

File: flow-server/src/main/java/com/vaadin/flow/server/communication/ServerRpcHandler.java (hunk 2 above)
Comment (caalador) [reply flow-server/src/main/java/com/vaadin/flow/server/communication/ServerRpcHandler.java]: changed

**Pull request merged**